File: LWHeroOC/Classes/IgnoreSubviewModifiersPreprocessor.m
//
//  IgnoreSubviewModifiersPreprocessor.m
//  Hero-ObjectiveC
//
//  Created by luca.li on 2017/1/24.
//  Copyright © 2017年 luca. All rights reserved.
//

#import "IgnoreSubviewModifiersPreprocessor.h"

@implementation IgnoreSubviewModifiersPreprocessor

- (void)processFromViews:(NSArray *)fromviews toViews:(NSArray *)toviews {
    [self processViews:fromviews];
    [self processViews:toviews];
}

- (void)processViews:(NSArray <UIView *> *)views {
    [views enumerateObjectsUsingBlock:^(UIView * _Nonnull view, NSUInteger viewIndex, BOOL * _Nonnull stop) {
        NSNumber *recursive = [self.context stateOfView:view].ignoreSubviewModifiers;
        if (recursive) {
            UIView *parentView = view;
            if ([view isKindOfClass:[UITableView class]]) {
                UIView *wrapperView = [view.subviews firstObject];
                if (wrapperView) {
                    parentView = wrapperView;
                }
            }
            
            if ([recursive boolValue]) {
                for (NSInteger i = viewIndex + 1; i < views.count; i ++) {
                    UIView *childView = views[i];
                    if (childView.superview == view.superview) {
                        break;
                    }
                    [self.context setState:nil toView:childView];
                }
            } else {
                for (UIView *subview in parentView.subviews) {
                    [self.context setState:nil toView:subview];
                }
            }
        }
    }];
}

@end

File: LWHero_swift/LWHero/Classes/IgnoreSubviewModifiersPreprocessor.swift
//
//  IgnoreSubviewModifiersPreprocessor.swift
//  LWHero
//
//  Created by Swift Port
//  Copyright © 2025. All rights reserved.
//

import Foundation
import UIKit

public class IgnoreSubviewModifiersPreprocessor: BasePreprocessor {
    // MARK: - HeroPreprocessor

    public override func process(fromViews: [UIView], toViews: [UIView]) {
        guard let context = context else { return }

        let allViews = fromViews + toViews

        for view in allViews {
            guard let state = context.state(of: view),
                  let ignoreSubviewModifiers = state.ignoreSubviewModifiers,
                  ignoreSubviewModifiers else { continue }

            // Remove modifiers from subviews
            processSubviews(of: view, recursive: ignoreSubviewModifiers)
        }
    }

    private func processSubviews(of view: UIView, recursive: Bool) {
        guard let context = context else { return }

        for subview in view.subviews {
            // Clear modifiers
            context.setState(HeroTargetState(), for: subview)

            if recursive {
                processSubviews(of: subview, recursive: true)
            }
        }
    }
}
